lib/transmuxer.js

```diff
@@ -1114,6 +1103,75 @@ calculateTrackBaseMediaDecodeTime = function (track) {
 };
 
 /**
+ * A Stream that can combine multiple streams (ie. audio & video)
+ * into a single output segment for MSE. Also supports audio-only
+ * and video-only streams.
+ */
+CoalesceStream = function(transmuxer) {
+  // Number of Tracks per output segment
+  // If greater than 1, we combine multiple
+  // tracks into a single segment
+  this.numberOfTracks = 0;
+
+  this.pendingTracks = [];
+  this.pendingBoxes = [];
+  this.pendingBytes = 0;
+
+  CoalesceStream.prototype.init.call(this);
+
+  // Take output from multiple
+  this.push = function(output) {
+    // append the media data to an init segment
+    var
+      offset = 0,
+      initSegment,
+      segmentData,
+      segmentType = output.track.type;
+
+    if (this.numberOfTracks > 1) {
+      segmentType = 'combined';
+    }
+
+    // Add this track to the list of pending tracks and store
+    // important information required for the construction of
+    // the final segment
+    this.pendingTracks.push(output.track);
+    this.pendingBoxes.push(output.boxes);
+    this.pendingBytes += output.boxes.byteLength;
+
+    // Once we have enough tracks from the various part of the
+    // re-muxing streams we can assemble the final output
+    if (this.pendingTracks.length >= this.numberOfTracks) {
+      initSegment = muxjs.mp4.initSegment(this.pendingTracks);
+
+      // Create a new typed array large enough to hold the init
+      // segment and all tracks
+      segmentData = new Uint8Array(initSegment.byteLength +
+                                   this.pendingBytes);
+
+      // Create an init segment containing a moov
+      // and track definitions
+      segmentData.set(initSegment);
+      offset += initSegment.byteLength;
+
+      // Append each moof (one per track) after the init segment
+      for (var i = 0; i < this.pendingBoxes.length; i++) {
+        segmentData.set(this.pendingBoxes[i], offset);
+        offset += this.pendingBoxes[i].byteLength;
+      }
+
+      // Emit the final segment via the transmuxer
+      transmuxer.trigger('data', {
```

Comment:
Could we have the transmuxer just listen for events on the coalesce stream and bubble them? Having this stream be aware of the transmuxer is inconsistent with the design of all the lower-level streams.